db_test.go

```diff
@@ -781,6 +783,71 @@ func TestTombstoneClean(t *testing.T) {
 	}
 }
 
+func TestTombstoneCleanFail(t *testing.T) {
+
+	db, close := openTestDB(t, nil)
+	defer close()
+
+	var expectedBlockDirs []string
+
+	// Create 2 empty blocks with some fake tombstones to have something to work with.
+	for i := 0; i < 2; i++ {
+		entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
+		uid := ulid.MustNew(ulid.Now(), entropy)
+		meta := &BlockMeta{
+			Version: 2,
+			ULID:    uid,
+		}
+		blockDir := filepath.Join(db.Dir(), uid.String())
+		block := createEmptyBlock(t, blockDir, meta)
+		tomb := memTombstones{}
+		tomb[0] = Intervals{{0, 1}}
+		block.tombstones = tomb
+
+		db.blocks = append(db.blocks, block)
+		expectedBlockDirs = append(expectedBlockDirs, blockDir)
+	}
+	db.compactor = &mockCompactorFailing{t: t}
+
+	// This should fail as we are using the failing compactor!
+	testutil.NotOk(t, db.CleanTombstones())
+
+	actualBlockDirs, err := blockDirs(db.dir)
```

Comment:
Before checking if actual blocks are as expected, should we also check if after a failed compaction, the actual blocks are not as expected? Only then we can verify that the new change is working. (Something like testing if `mockCompactorFailing` is working as expected. We can do that check before returning an error in `Write` I guess)

Reply:
yep , I thought about this as well, but can't see how we can implement this nicely. Any  idea?

Reply:
We can store the `expectedBlockDirs` in `mockCompactorFailing`. And just before returning the error in `Write`, we can check for not equal in the similar way how we check for equal.

I don't know how we can do that inside the actual test function. So this is a small hack around as you are already storing `t  *testing.T` object inside `mockCompactorFailing`

Reply:
ooh , yeah that is a good idea , Will try it now.